include/deal.II/matrix_free/operators.h

```diff
@@ -1802,17 +1803,37 @@ namespace MatrixFreeOperators
     VectorType &diagonal_vector = this->diagonal_entries->get_vector();
     this->initialize_dof_vector(inverse_diagonal_vector);
     this->initialize_dof_vector(diagonal_vector);
-    inverse_diagonal_vector = Number(1.);
-    apply_add(diagonal_vector, inverse_diagonal_vector);
 
-    this->set_constrained_entries_to_one(diagonal_vector);
+    auto diagonal_evaluation = [](auto &integrator) {
+      integrator.evaluate(EvaluationFlags::values);
+      for (unsigned int q = 0; q < integrator.n_q_points; ++q)
+        integrator.submit_value(integrator.get_value(q), q);
+      integrator.integrate(EvaluationFlags::values);
+    };
+
+    std::function<void(
+      FEEvaluation<
+        dim,
+        fe_degree,
+        n_q_points_1d,
+        n_components,
+        typename Base<dim, VectorType, VectorizedArrayType>::value_type,
+        VectorizedArrayType> &)>
+      diagonal_evaluation_f(diagonal_evaluation);
```

Comment:
You don't need to wrap this function into lambda.

Reply:
do you mean 'wrap into a `std::function`? I tried not adding the cast explicitly myself and my compiler wasn't able to resolve the types.

Reply:
There is a second version of the function that take a member function directly.

Reply:
Ah. Would you prefer I make the lambda a member function?

Reply:
Oh sorry, I thought there is already such a member function. If that is not the case let's keep it like this.

Reply:
There is but it doesn't conform to the interface that `compute_diagonal` wants.